File: apps/primitive-compiler/src/#syntax/tokenizer.ts
import { Span, Token, TokenKind } from './common/index.js';

export class Tokenizer {
  private index = 0n;
  constructor(private source: string) {}

  *[Symbol.iterator](): Iterator<Token, void> {
    while (true) {
      const token = this.next();
      if (token == null) break;
      yield token;
    }
  }

  next(): Token | null {
    return (
      this.nextVerticalSpace() ??
      this.nextHorizontalSpace() ??
      this.nextIdentifierOrKeyword() ??
      this.nextPunctuation()
    );
  }

  private peek(): string | null {
    return this.index < this.source.length
      ? this.source[this.index.toString() as unknown as number]
      : null;
  }

  HORIZONTAL_SPACE = [
    '\t', // CHARACTER TABULATION
    ' ', // SPACE
    '\u{00AD}', // SOFT HYPHEN
    '\u{00A0}', // NO-BREAK SPACE
    '\u{1680}', // OGHAM SPACE MARK
    '\u{2000}', // EN QUAD
    '\u{2001}', // EM QUAD
    '\u{2002}', // EN SPACE
    '\u{2003}', // EM SPACE
    '\u{2004}', // THREE-PER-EM SPACE
    '\u{2005}', // FOUR-PER-EM SPACE
    '\u{2006}', // SIX-PER-EM SPACE
    '\u{2007}', // FIGURE SPACE
    '\u{2008}', // PUNCTUATION SPACE
    '\u{2009}', // THIN SPACE
    '\u{200A}', // HAIR SPACE
    '\u{200B}', // ZERO WIDTH SPACE
    '\u{200E}', // LEFT-TO-RIGHT MARK
    '\u{200F}', // RIGHT-TO-LEFT MARK
    '\u{202F}', // NARROW NO-BREAK SPACE
    '\u{205F}', // MEDIUM MATHEMATICAL SPACE
    '\u{3000}', // IDEPGRAPHIC SPACE
    '\u{FEFF}', // ZERO WIDTH NO-BREAK SPACE
  ];
  private nextHorizontalSpace(): Token | null {
    const begin = this.index;

    let content = '';

    let current: string | null;
    do {
      current = this.peek();
      if (current == null || !this.HORIZONTAL_SPACE.includes(current)) break;

      content += current;
      this.index++;
    } while (true);

    if (content.length === 0) return null;

    return Token.make({
      content,
      kind: TokenKind.SpaceHorizotanl,
      span: Span.make({ begin, end: this.index }),
    });
  }

  VERTICAL_SPACE = [
    '\n', // LINE FEED
    '\u{000B}', // LINE TABULATION
    '\u{000C}', // FORM FEED
    '\r', // CARRIAGE RETURN
    '\u{0085}', // NEXT LINE
    '\u{2028}', // LINE SEPARATOR
    '\u{2029}', // PARAGRAPH SEPARATOR
  ];
  private nextVerticalSpace(): Token | null {
    const begin = this.index;

    let content = this.peek();

    if (content == null || !this.VERTICAL_SPACE.includes(content)) return null;
    this.index++;

    if (content === '\r' && this.peek() === '\n') {
      this.index++;

      return Token.make({
        content: '\r\n',
        kind: TokenKind.SpaceVertical,
        span: Span.make({ begin, end: this.index }),
      });
    } else {
      return Token.make({
        content,
        kind: TokenKind.SpaceVertical,
        span: Span.make({ begin, end: this.index }),
      });
    }
  }

  ID_Start = /^\p{ID_Start}$/u;
  ID_Continue = /^\p{ID_Continue}$/u;
  private nextIdentifierOrKeyword(): Token | null {
    const begin = this.index;

    let content = this.peek();
    if (content == null || !this.ID_Start.test(content)) return null;
    this.index++;

    let current: string | null;
    while ((current = this.peek()) != null && this.ID_Continue.test(current)) {
      content += current;
      this.index++;
    }

    let kind: TokenKind =
      {
        enum: TokenKind.KeywordEnum,
        fn: TokenKind.KeywordFn,
        mut: TokenKind.KeywordMut,
        struct: TokenKind.KeywordStruct,
      }[content] ?? TokenKind.Identifier;

    return Token.make({
      content,
      kind,
      span: Span.make({ begin, end: this.index }),
    });
  }

  PunctuationsTrieRoot: TrieNode = {
    children: {
      '!': { conclusion: TokenKind.PunctuationExclamationMark },
      '#': { conclusion: TokenKind.PunctuationNumberSign },
      $: { conclusion: TokenKind.PunctuationDollarSign },
      '%': { conclusion: TokenKind.PunctuationPercentSign },
      '&': { conclusion: TokenKind.PunctuationAmpersand },
      '*': { conclusion: TokenKind.PunctuationAsterisk },
      '+': { conclusion: TokenKind.PunctuationPlusSign },
      ',': { conclusion: TokenKind.PunctuationComma },
      '-': { conclusion: TokenKind.PunctuationHyphenMinus },
      '.': { conclusion: TokenKind.PunctuationFullStop },
      '/': { conclusion: TokenKind.PunctuationSolidus },
      ':': { conclusion: TokenKind.PunctuationColon },
      ';': { conclusion: TokenKind.PunctuationSemicolon },
      '<': { conclusion: TokenKind.PunctuationLessThanSign },
      '=': {
        conclusion: TokenKind.PunctuationEqualsSign,
        children: {
          '>': { conclusion: TokenKind.PunctuationsFatArrow },
        },
      },
      '>': { conclusion: TokenKind.PunctuationGreaterThanSign },
      '?': { conclusion: TokenKind.PunctuationQuestionMark },
      '@': { conclusion: TokenKind.PunctuationCommercialAt },
      '\\': { conclusion: TokenKind.PunctuationReverseSolidus },
      '^': { conclusion: TokenKind.PunctuationCircumflexAccent },
      '|': { conclusion: TokenKind.PunctuationVerticalLine },
      '~': { conclusion: TokenKind.PunctuationTilde },
      '(': { conclusion: TokenKind.PunctuationLeftParenthesis },
      '[': { conclusion: TokenKind.PunctuationLeftSquareBracket },
      '{': { conclusion: TokenKind.PunctuationLeftCurlyBracket },
      ')': { conclusion: TokenKind.PunctuationRightParenthesis },
      ']': { conclusion: TokenKind.PunctuationRightSquareBracket },
      '}': { conclusion: TokenKind.PunctuationRightCurlyBracket },
    },
  };
  nextPunctuation(): Token | null {
    const begin = this.index;

    let lastSuccess: Token | null = null;
    let node = this.PunctuationsTrieRoot;
    let content = '';
    let current: string | null;
    while (
      (current = this.peek()) != null &&
      node.children != null &&
      current in node.children
    ) {
      this.index++;
      node = node.children[current];
      content += current;
      if (node.conclusion != null) {
        lastSuccess = Token.make({
          kind: node.conclusion,
          content,
          span: Span.make({
            begin,
            end: this.index,
          }),
        });
      }
    }

    if (lastSuccess != null) {
      this.index = lastSuccess.span.end;
    }

    return lastSuccess;
  }
}

interface TrieNode {
  conclusion?: TokenKind;
  children?: Record<string, TrieNode>;
}

File: crates/lumo-syntax/src/tokenizer.rs
use std::iter::once;

use lumo_core::{Offset, Span, Spanned, Token, TokenKind};
use unicode_ident::{is_xid_continue, is_xid_start};
use winnow::{
    Parser, Result, Stateful,
    combinator::{alt, dispatch, empty, fail, repeat},
    error::{ContextError, ParseError},
    token::any,
};

type Input<'a> = Stateful<&'a str, Offset>;

pub fn tokenize(source: &str) -> Result<Vec<Spanned<Token>>, ParseError<Input, ContextError>> {
    token_list.parse(Input {
        state: Offset {
            offset: 0,
            line: 1,
            col: 0,
        },
        input: source,
    })
}

pub fn token_list(i: &mut Input) -> Result<Vec<Spanned<Token>>> {
    let mut result: Vec<_> = repeat(0.., token).parse_next(i)?;
    result.push(make_token(TokenKind::Eof, "", i));
    Ok(result)
}

fn token(i: &mut Input) -> Result<Spanned<Token>> {
    alt((
        space_horizontal,
        space_vertical,
        ident_keyword,
        punctuations,
    ))
    .parse_next(i)
}

fn space_horizontal(i: &mut Input) -> Result<Spanned<Token>> {
    let content: Vec<_> = repeat(
        1..,
        alt((
            alt((
                "\t",       // CHARACTER TABULATION
                " ",        // SPACE
                "\u{00AD}", // SOFT HYPHEN
                "\u{00A0}", // NO-BREAK SPACE
                "\u{1680}", // OGHAM SPACE MARK
                "\u{2000}", // EN QUAD
                "\u{2001}", // EM QUAD
                "\u{2002}", // EN SPACE
                "\u{2003}", // EM SPACE
                "\u{2004}", // THREE-PER-EM SPACE
                "\u{2005}", // FOUR-PER-EM SPACE
                "\u{2006}", // SIX-PER-EM SPACE
                "\u{2007}", // FIGURE SPACE
                "\u{2008}", // PUNCTUATION SPACE
                "\u{2009}", // THIN SPACE
                "\u{200A}", // HAIR SPACE
                "\u{200B}", // ZERO WIDTH SPACE
                "\u{200E}", // LEFT-TO-RIGHT MARK
                "\u{200F}", // RIGHT-TO-LEFT MARK
                "\u{202F}", // NARROW NO-BREAK SPACE
                "\u{205F}", // MEDIUM MATHEMATICAL SPACE
            )),
            alt((
                "\u{3000}", // IDEPGRAPHIC SPACE
                "\u{FEFF}", // ZERO WIDTH NO-BREAK SPACE
            )),
        )),
    )
    .parse_next(i)?;
    Ok(make_token(TokenKind::SpaceHorizontal, content.join(""), i))
}

fn space_vertical(i: &mut Input) -> Result<Spanned<Token>> {
    let content = alt((
        "\r\n",     // CRLF
        "\n",       // LINE FEED
        "\u{000B}", // LINE TABULATION
        "\u{000C}", // FORM FEED
        "\r",       // CARRIAGE RETURN
        "\u{0085}", // NEXT LINE
        "\u{2028}", // LINE SEPARATOR
        "\u{2029}", // PARAGRAPH SEPARATOR
    ))
    .parse_next(i)?;
    Ok(make_token(TokenKind::SpaceVertical, content, i))
}

fn ident_keyword(i: &mut Input) -> Result<Spanned<Token>> {
    let (start, cont): (_, Vec<_>) = (
        any.verify(|c| *c == '_' || is_xid_start(*c)),
        repeat(0.., any.verify(|c| is_xid_continue(*c))),
    )
        .parse_next(i)?;
    let content = once(start).chain(cont).collect::<String>();

    Ok(make_token(
        match content.as_ref() {
            "enum" => TokenKind::KeywordEnum,
            "fn" => TokenKind::KeywordFn,
            "let" => TokenKind::KeywordLet,
            "match" => TokenKind::KeywordMatch,
            "mut" => TokenKind::KeywordMut,
            "struct" => TokenKind::KeywordStruct,
            "_" => TokenKind::IdentifierUnderscore,
            _ => TokenKind::IdentifierIdentifier,
        },
        content,
        i,
    ))
}

fn punctuations(i: &mut Input) -> Result<Spanned<Token>> {
    let (kind, content) = dispatch! {any;
        '!' => empty.value(TokenKind::PunctuationExclamationMark),
        '#' => empty.value(TokenKind::PunctuationNumberSign),
        '$' => empty.value(TokenKind::PunctuationDollarSign),
        '%' => empty.value(TokenKind::PunctuationPercentSign),
        '&' => empty.value(TokenKind::PunctuationAmpersand),
        '*' => empty.value(TokenKind::PunctuationAsterisk),
        '+' => empty.value(TokenKind::PunctuationPlusSign),
        ',' => empty.value(TokenKind::PunctuationComma),
        '-' => empty.value(TokenKind::PunctuationHyphenMinus),
        '.' => empty.value(TokenKind::PunctuationFullStop),
        '/' => empty.value(TokenKind::PunctuationSolidus),
        ':' => empty.value(TokenKind::PunctuationColon),
        ';' => empty.value(TokenKind::PunctuationSemicolon),
        '<' => empty.value(TokenKind::PunctuationLessThanSign),
        '=' => alt((
            '>'.value(TokenKind::PunctuationsFatArrow),
            empty.value(TokenKind::PunctuationEqualsSign),
        )),
        '>' => empty.value(TokenKind::PunctuationGreaterThanSign),
        '?' => empty.value(TokenKind::PunctuationQuestionMark),
        '@' => empty.value(TokenKind::PunctuationCommercialAt),
        '\\' => empty.value(TokenKind::PunctuationReverseSolidus),
        '^' => empty.value(TokenKind::PunctuationCircumflexAccent),
        '|' => empty.value(TokenKind::PunctuationVerticalLine),
        '~' => empty.value(TokenKind::PunctuationTilde),
        '(' => empty.value(TokenKind::PunctuationLeftParenthesis),
        '[' => empty.value(TokenKind::PunctuationLeftSquareBracket),
        '{' => empty.value(TokenKind::PunctuationLeftCurlyBracket),
        ')' => empty.value(TokenKind::PunctuationRightParenthesis),
        ']' => empty.value(TokenKind::PunctuationRightSquareBracket),
        '}' => empty.value(TokenKind::PunctuationRightCurlyBracket),
        _ => fail,
    }
    .with_taken()
    .parse_next(i)?;
    Ok(make_token(kind, content, i))
}

fn make_token(kind: TokenKind, content: impl AsRef<str>, i: &mut Input) -> Spanned<Token> {
    let content = content.as_ref().to_owned();
    let begin = i.state.clone();
    i.state.offset += content.len();
    if kind == TokenKind::SpaceVertical {
        i.state.col = 0;
        i.state.line += 1;
    } else {
        i.state.col += content.len();
    }
    Spanned::new(Span::new(begin, i.state.clone()), Token { kind, content })
}
